Subject: Disable i386 APIC by default
From: ak@suse.de

Don't enable local APIC unless the "apic" option is specified on i386
UP kernels.

We also now enable APIC for systems newer than 2001 or anything
that has multiple processors

Index: linux-2.6.16/Documentation/kernel-parameters.txt
===================================================================
--- linux-2.6.16.orig/Documentation/kernel-parameters.txt
+++ linux-2.6.16/Documentation/kernel-parameters.txt
@@ -235,6 +235,12 @@ running once the system is up.
 	apm=		[APM] Advanced Power Management
 			See header of arch/i386/kernel/apm.c.
 
+	apic		[UP,APIC,i386] Tells the kernel to make use of the APIC on
+	                uniprocessor systems. When CONFIG_X86_APIC_OFF is set it 
+			needs this parameter to really activate it (needed to 
+			avoid problems with certain machines). This enables IO 
+			and local APICs.
+
 	applicom=	[HW]
 			Format: <mem>,<irq>
 
Index: linux-2.6.16/arch/i386/kernel/apic.c
===================================================================
--- linux-2.6.16.orig/arch/i386/kernel/apic.c
+++ linux-2.6.16/arch/i386/kernel/apic.c
@@ -27,6 +27,7 @@
 #include <linux/sysdev.h>
 #include <linux/cpu.h>
 #include <linux/module.h>
+#include <linux/dmi.h>
 
 #include <asm/atomic.h>
 #include <asm/smp.h>
@@ -743,7 +744,11 @@ static void apic_pm_activate(void) { }
 
 static int __init apic_set_verbosity(char *str)
 {
-	if (strcmp("debug", str) == 0)
+	if (*str == '=') 
+		str++;
+	if (*str == 0) 
+		;/* parsed early already */
+	else if (strcmp("debug", str) == 0)
 		apic_verbosity = APIC_DEBUG;
 	else if (strcmp("verbose", str) == 0)
 		apic_verbosity = APIC_VERBOSE;
@@ -754,15 +759,16 @@ static int __init apic_set_verbosity(cha
 	return 0;
 }
 
-__setup("apic=", apic_set_verbosity);
+__setup("apic", apic_set_verbosity);
 
 static int __init detect_init_APIC (void)
 {
 	u32 h, l, features;
-
+#ifdef CONFIG_X86_APIC_OFF
 	/* Disabled by kernel option? */
 	if (enable_local_apic < 0)
 		return -1;
+#endif
 
 	switch (boot_cpu_data.x86_vendor) {
 	case X86_VENDOR_AMD:
@@ -851,8 +857,6 @@ void __init init_apic_mappings(void)
 		apic_phys = mp_lapic_addr;
 
 	set_fixmap_nocache(FIX_APIC_BASE, apic_phys);
-	printk(KERN_DEBUG "mapped APIC to %08lx (%08lx)\n", APIC_BASE,
-	       apic_phys);
 
 	/*
 	 * Fetch the APIC ID of the BSP in case we have a
@@ -885,8 +889,6 @@ fake_ioapic_page:
 				ioapic_phys = __pa(ioapic_phys);
 			}
 			set_fixmap_nocache(idx, ioapic_phys);
-			printk(KERN_DEBUG "mapped IOAPIC to %08lx (%08lx)\n",
-			       __fix_to_virt(idx), ioapic_phys);
 			idx++;
 		}
 	}
@@ -1309,17 +1311,101 @@ fastcall void smp_error_interrupt(struct
 	irq_exit();
 }
 
+static int __init need_apic(struct dmi_system_id *d)
+{ 
+#ifdef CONFIG_X86_LOCAL_APIC
+	extern int enable_local_apic;
+	if (enable_local_apic == 0)  {
+		enable_local_apic = 1;
+		printk(
+KERN_INFO "%s detected. Enabling local APIC. Overwrite with \"nolapic\"\n",
+		       d->ident);
+#endif
+	}
+	return 0;
+} 
+
+/* The table is mostly covered by the generic checks below anyways, but keep it for now */ 
+static struct dmi_system_id __initdata apic_dmi_table[] = {
+	/* Systems that need APIC */
+	/* Multinode Summit systems need APIC to boot */
+	{ need_apic, "IBM x445", 
+	  { DMI_MATCH(DMI_SYS_VENDOR, "IBM"),
+	    DMI_MATCH(DMI_PRODUCT_NAME, "xSeries 440") }},
+	{ need_apic, "IBM x445", 
+	  { DMI_MATCH(DMI_SYS_VENDOR, "IBM"),
+	    DMI_MATCH(DMI_PRODUCT_NAME, "xSeries 445") }},
+	{ need_apic, "IBM x460",
+	  { DMI_MATCH(DMI_SYS_VENDOR, "IBM"),
+	    DMI_MATCH(DMI_PRODUCT_NAME, "eserver xSeries 460")}},
+	{}
+};
+
+#ifdef CONFIG_X86_APIC_OFF
+static __init int dmi_enable_apic(void)
+{
+	int year;
+	char *vendor; 
+
+	/* If the machine has more than one CPU try to use APIC because it'll 
+	   be running the SMP kernel with APIC soon anyways. 
+	   This won't cover dual core, but they are handled by the date check 
+	   below. */
+	if (dmi_num_cpus > 1)
+		return 1;
+
+	year = dmi_get_year(DMI_BIOS_DATE);
+	vendor = dmi_get_system_info(DMI_BIOS_VENDOR);
+
+	/* All Intel BIOS since 1998 assumed APIC on. Don't include 1998 itself
+	   because we're not sure for that. */
+	if (vendor && !strncmp(vendor, "Intel", 5))
+		return year > 1998;
+
+	/* Use APIC for anything since 2001 */
+	return year >= 2001;
+}
+
+void __init dmi_check_apic(void)
+{
+	if (enable_local_apic != 0)
+		return;
+	if (!dmi_enable_apic()) {
+		enable_local_apic = -1;
+		clear_bit(X86_FEATURE_APIC, boot_cpu_data.x86_capability);
+		printk(
+	KERN_INFO "IO/L-APIC disabled because your old system seems to be old\n"); 
+		printk(KERN_INFO "overwrite with \"apic\"\n");
+	} else {
+		printk(
+	KERN_INFO "IO/L-APIC allowed because system is MP or new enough\n");
+	}
+}
+#endif
+
 /*
  * This initializes the IO-APIC and APIC hardware if this is
  * a UP kernel.
  */
 int __init APIC_init_uniprocessor (void)
 {
+	dmi_check_system(apic_dmi_table);
+
+#ifdef CONFIG_X86_APIC_OFF
+	if (enable_local_apic < 0) { 
+		clear_bit(X86_FEATURE_APIC, boot_cpu_data.x86_capability);
+		nr_ioapics = 0;
+		return -1;
+	}
+#endif
+
 	if (enable_local_apic < 0)
 		clear_bit(X86_FEATURE_APIC, boot_cpu_data.x86_capability);
 
-	if (!smp_found_config && !cpu_has_apic)
+	if (!smp_found_config && !cpu_has_apic) { 
+		nr_ioapics = 0;
 		return -1;
+	}
 
 	/*
 	 * Complain if the BIOS pretends there is one.
Index: linux-2.6.16/arch/i386/kernel/setup.c
===================================================================
--- linux-2.6.16.orig/arch/i386/kernel/setup.c
+++ linux-2.6.16/arch/i386/kernel/setup.c
@@ -865,6 +865,10 @@ static void __init parse_cmdline_early (
 		/* disable IO-APIC */
 		else if (!memcmp(from, "noapic", 6))
 			disable_ioapic_setup();
+ 		/* SUSE specific option */
+ 		else if (!memcmp(from, "apic", 4) && 
+				(from[4] == ' ' || from[4] == '\0'))
+ 		        lapic_enable();
 #endif /* CONFIG_X86_IO_APIC */
 #endif /* CONFIG_ACPI */
 
@@ -1604,6 +1608,9 @@ void __init setup_arch(char **cmdline_p)
 
 
 	dmi_scan_machine();
+#ifdef CONFIG_X86_APIC_OFF
+	dmi_check_apic();
+#endif
 
 #ifdef CONFIG_X86_GENERICARCH
 	generic_apic_probe(*cmdline_p);
Index: linux-2.6.16/arch/i386/Kconfig
===================================================================
--- linux-2.6.16.orig/arch/i386/Kconfig
+++ linux-2.6.16/arch/i386/Kconfig
@@ -241,6 +241,15 @@ config X86_UP_APIC
 	  performance counters), and the NMI watchdog which detects hard
 	  lockups.
 
+config X86_APIC_OFF
+        bool "Disable local/IO APIC by default" 
+	depends on X86_LOCAL_APIC  && !SMP
+	default n
+	help
+	  When this option is enabled the IO/local APIC code is compiled in, but 
+	  only enabled when the kernel is booted with "apic" on the kernel 
+	  command line  or an DMI entry forced APIC mode. 
+
 config X86_UP_IOAPIC
 	bool "IO-APIC support on uniprocessors"
 	depends on X86_UP_APIC
Index: linux-2.6.16/arch/x86_64/kernel/apic.c
===================================================================
--- linux-2.6.16.orig/arch/x86_64/kernel/apic.c
+++ linux-2.6.16/arch/x86_64/kernel/apic.c
@@ -43,6 +43,9 @@ int apic_runs_main_timer;
 int apic_calibrate_pmtmr __initdata;
 
 int disable_apic_timer __initdata;
+/* just used to communicate with shared i386 code: */
+int enable_local_apic = 1; 
+
 
 /*
  * cpu_mask that denotes the CPUs that needs timer interrupt coming in as
@@ -1110,6 +1113,7 @@ int __init APIC_init_uniprocessor (void)
 	}
 	if (!cpu_has_apic) { 
 		disable_apic = 1;
+		enable_local_apic = -1; 
 		printk(KERN_INFO "Apic disabled by BIOS\n");
 		return -1;
 	}
@@ -1136,12 +1140,14 @@ int __init APIC_init_uniprocessor (void)
 
 static __init int setup_disableapic(char *str) 
 { 
+	enable_local_apic = -1;
 	disable_apic = 1;
 	return 0;
 } 
 
 static __init int setup_nolapic(char *str) 
 { 
+	enable_local_apic = -1;
 	disable_apic = 1;
 	return 0;
 } 
Index: linux-2.6.16/include/asm-i386/mach-default/mach_apic.h
===================================================================
--- linux-2.6.16.orig/include/asm-i386/mach-default/mach_apic.h
+++ linux-2.6.16/include/asm-i386/mach-default/mach_apic.h
@@ -56,8 +56,10 @@ static inline physid_mask_t ioapic_phys_
 
 static inline void clustered_apic_check(void)
 {
+#ifdef CONFIG_SMP
 	printk("Enabling APIC mode:  %s.  Using %d I/O APICs\n",
 					"Flat", nr_ioapics);
+#endif
 }
 
 static inline int multi_timer_check(int apic, int irq)
Index: linux-2.6.16/arch/i386/kernel/dmi_scan.c
===================================================================
--- linux-2.6.16.orig/arch/i386/kernel/dmi_scan.c
+++ linux-2.6.16/arch/i386/kernel/dmi_scan.c
@@ -6,6 +6,8 @@
 #include <linux/bootmem.h>
 #include <linux/slab.h>
 
+int dmi_num_cpus;
+
 static char * __init dmi_string(struct dmi_header *dm, u8 s)
 {
 	u8 *bp = ((u8 *) dm) + dm->length;
@@ -176,6 +178,9 @@ static void __init dmi_decode(struct dmi
 		dmi_save_ident(dm, DMI_BOARD_NAME, 5);
 		dmi_save_ident(dm, DMI_BOARD_VERSION, 6);
 		break;
+	case 4: 	/* Central Processor */
+		dmi_num_cpus++;
+		break;
 	case 10:	/* Onboard Devices Information */
 		dmi_save_devices(dm);
 		break;
@@ -299,3 +304,31 @@ struct dmi_device * dmi_find_device(int 
 	return NULL;
 }
 EXPORT_SYMBOL(dmi_find_device);
+
+/**
+ *	dmi_get_year - Return year of a DMI date
+ *	@field:	data index (like dmi_get_system_info)
+ *
+ *	Handles y2k issues.
+ */
+int dmi_get_year(int field)
+{
+	int year;
+	char *s = dmi_get_system_info(field);
+
+	if (!s || !*s)
+		return 0;
+	s = strrchr(s, '/');
+	if (!s)
+		return 0;
+
+	s += 1;
+	year = simple_strtoul(s, NULL, 0);
+	if (year < 100) {	/* 2-digit year */
+		year += 1900;
+		if (year < 1996)	/* no dates < spec 1.0 */
+			year += 100;
+	}
+
+	return year;
+}
Index: linux-2.6.16/include/linux/dmi.h
===================================================================
--- linux-2.6.16.orig/include/linux/dmi.h
+++ linux-2.6.16/include/linux/dmi.h
@@ -69,12 +69,16 @@ extern struct dmi_device * dmi_find_devi
 	struct dmi_device *from);
 extern void dmi_scan_machine(void);
 
+extern int dmi_num_cpus;
+extern int dmi_get_year(int field);
+
 #else
 
 static inline int dmi_check_system(struct dmi_system_id *list) { return 0; }
 static inline char * dmi_get_system_info(int field) { return NULL; }
 static inline struct dmi_device * dmi_find_device(int type, const char *name,
 	struct dmi_device *from) { return NULL; }
+static inline int dmi_get_year(int year) { return 0; }
 
 #endif
 
Index: linux-2.6.16/include/asm-i386/apic.h
===================================================================
--- linux-2.6.16.orig/include/asm-i386/apic.h
+++ linux-2.6.16/include/asm-i386/apic.h
@@ -139,6 +139,8 @@ void switch_ipi_to_APIC_timer(void *cpum
 
 extern int timer_over_8254;
 
+extern void dmi_check_apic(void);
+
 #else /* !CONFIG_X86_LOCAL_APIC */
 static inline void lapic_shutdown(void) { }
 
